### PR TITLE
MCOL 641 Combined PR for commits 2024e8f and c7dbeb1

### PR DESCRIPTION
**Pull request by tntnatbry**

### CODE REVIEW

**Comment by tntnatbry** [comment]
This PR does not yet address your comments in https://github.com/drrtuy/mariadb-columnstore-engine/pull/10.

File: dbcon/joblist/jlf_common.cpp [hunk 1]
```diff
@@ -332,7 +332,7 @@ string extractTableAlias(const SSC& sc)
 //------------------------------------------------------------------------------
 CalpontSystemCatalog::OID isDictCol(const CalpontSystemCatalog::ColType& colType)
 {
-    if (colType.colDataType == CalpontSystemCatalog::BINARY) return 0;
+    if (utils::isWideDecimalType(colType)) return 0;
```
Comment (drrtuy) [inline dbcon/joblist/jlf_common.cpp]: Hm, I don't think that we want to treat any BINARY as a dictionary.

File: dbcon/joblist/jlf_execplantojoblist.cpp [hunk 2]
```diff
@@ -450,8 +456,10 @@ int64_t convertValueNum(const string& str, const CalpontSystemCatalog::ColType&
 #else
                 v = boost::any_cast<int32_t>(anyVal);
 #endif
-            else
+            else if (ct.colWidth == execplan::CalpontSystemCatalog::EIGHT_BYTE)
```
Comment (drrtuy) [inline dbcon/joblist/jlf_execplantojoblist.cpp]: I think people tend to prefer DECIMAL(38)  over all others so we want to put it first and use LIKELY despite the fact it won't give much gain.

Comment (drrtuy) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: And I would sort the if predicates in the opposite order.

File: utils/common/columnwidth.h [hunk 3]
```diff
@@ -35,6 +36,13 @@ namespace utils
         return width <= MAXLEGACYWIDTH;
     }
 
+    inline bool isWideDecimalType(const execplan::CalpontSystemCatalog::ColType& ct)
+    {
+        return ((ct.colDataType == execplan::CalpontSystemCatalog::DECIMAL ||
+            ct.colDataType == execplan::CalpontSystemCatalog::UDECIMAL) &&
+            ct.colWidth == MAXCOLUMNWIDTH);
```
Comment (drrtuy) [inline utils/common/columnwidth.h]: Imagine we increase the MAXCOLUMNWIDTH. This must be a dedicated constant.

Comment (drrtuy) [reply utils/common/columnwidth.h]: But I like the idea.

File: writeengine/wrapper/writeengine.cpp [hunk 4]
```diff
@@ -70,6 +70,9 @@ namespace WriteEngine
 {
 StopWatch timer;
 
+using dataconvert::int128_t;
+using dataconvert::uint128_t;
```
Comment (drrtuy) [inline writeengine/wrapper/writeengine.cpp]: This is an incorrect approach. There shouldn't be these definitions in dataconvert namespace.

Comment (drrtuy) [reply writeengine/wrapper/writeengine.cpp]: They are redundant.

File: writeengine/bulk/we_bulkloadbuffer.cpp [hunk 5]
```diff
@@ -994,7 +994,9 @@ void BulkLoadBuffer::convert(char* field, int fieldLength,
                             }
                             else
                             {
-                                dataconvert::atoi128(string(field), bigllVal);
+                                bool saturate = false;
```
Comment (drrtuy) [inline writeengine/bulk/we_bulkloadbuffer.cpp]: This if-else should have wide-decimals block first with expressions put into LIKELY()

File: utils/rowgroup/rowgroup.cpp [hunk 6]
```diff
@@ -593,22 +593,23 @@ string Row::toString() const
                     os << " " << dec;
                     break;
                 }
+                // WIP MCOL-641
                 case CalpontSystemCatalog::BINARY:
                     std::cout << __FILE__<< ":" <<__LINE__ << " Fix for 16 Bytes ?" << std::endl;
```
Comment (drrtuy) [inline utils/rowgroup/rowgroup.cpp]: It's time to remove this.

File: utils/joiner/tuplejoiner.cpp [hunk 7]
```diff
@@ -1146,7 +1146,7 @@ void TupleJoiner::updateCPData(const Row& r)
                     }
                 }
             }
-            else if (utils::isWide(r.getColumnWidth(colIdx))
+            else if (r.getColumnWidth(colIdx) == utils::MAXCOLUMNWIDTH
```
Comment (drrtuy) [inline utils/joiner/tuplejoiner.cpp]: This depends on MAXCOLUMNWIDTH that can change w/o notifications.

File: dbcon/joblist/jlf_common.cpp (hunk 1 above)
Comment (tntnatbry) [reply dbcon/joblist/jlf_common.cpp]: I have added back the predicate for BINARY, not treating it as dictionary if it has narrow width (<=8).

File: dbcon/joblist/jlf_execplantojoblist.cpp (hunk 2 above)
Comment (tntnatbry) [reply dbcon/joblist/jlf_execplantojoblist.cpp]: Done.

File: writeengine/wrapper/writeengine.cpp (hunk 4 above)
Comment (tntnatbry) [reply writeengine/wrapper/writeengine.cpp]: Good catch, removed it.

File: writeengine/bulk/we_bulkloadbuffer.cpp (hunk 5 above)
Comment (tntnatbry) [reply writeengine/bulk/we_bulkloadbuffer.cpp]: Done.

File: utils/joiner/tuplejoiner.cpp (hunk 7 above)
Comment (tntnatbry) [reply utils/joiner/tuplejoiner.cpp]: Changed it to datatypes::MAXDECIMALWIDTH

**Pull request merged**